Specialize findprev for BitArrays

mbauman opened the pull request:
This is the first part of merging my refactor of IntSets to use BitArrays (ref. [IntSets.jl](https://github.com/mbauman/IntSets.jl) and #9679).

This adds a specialized `findprev` method for BitArrays (similar to `findnext`) that searches through the array 64 bits at a time.  Here's an exaggerated example of how much this can improve performance:

``` julia
julia> b = falses(10^10); b[10] = true;

julia> findprev(b,10^10); @time findprev(b,10^10) # Before:
elapsed time: 19.229230915 seconds (13824 bytes allocated)
10

julia> findprev(b,10^10); @time findprev(b,10^10) # After:
elapsed time: 0.265889115 seconds (96 bytes allocated)
10
```

Also adds and tests an unexported `findprevnot` (like `findnextnot`).

Secondly, in the course of writing the additional tests, I discovered that there was a bug in `findnext(::Function, ::BitArray, …)`. It did not check bounds or could return the wrong result if the function was always true or always false.  This fixes it and adds tests.

cc @carlobaldassi

mbauman committed:
Specialize findprev for BitArrays

This adds a specialized `findprev` method for BitArrays (similar to `findnext`) that searches through the array 64 bits at a time.  Here's an exaggerated example of how much this can improve performance:

```julia
julia> b = falses(10^10); b[10] = true;

julia> findprev(b,10^10); @time findprev(b,10^10) # Before:
elapsed time: 19.229230915 seconds (13824 bytes allocated)
10

julia> findprev(b,10^10); @time findprev(b,10^10) # After:
elapsed time: 0.265889115 seconds (96 bytes allocated)
10
```

Also adds and tests an unexported `findprevnot` (like `findnextnot`), which will be useful in converting IntSets to use BitArrays.

mbauman committed:
Fix bug in `findnext(::Function, ::BitArray, ...)`

Previously, `findnext(::Function, ::BitArray, ...)` did not check bounds or could return the wrong result if the function was always true or always false.  This fixes it and adds tests.

timholy commented:
This is pretty awesome. I don't have time ATM to check the code, and @carlobaldassi would be much better anyway. But the extensive test additions are obviously welcome.

carlobaldassi commented on base/bitarray.jl:
```diff
@@ -1377,14 +1377,90 @@ end
+
+    start > 0 || throw(BoundsError(B, start))
+    start > length(B) && return 0
+    length(B) == 0 && return 0
```
this seems redundant with the previous two checks

mbauman replied:
Yup, sure enough.

carlobaldassi commented on base/bitarray.jl:
```diff
@@ -1377,14 +1377,90 @@ end
+
+    Bc = B.chunks
+
+    chunk_start = _div64(start)+1
```
should be `_div64(start-1)+1`. Test for example `findprev(trues(64), 64)`.

carlobaldassi replied:
(You need to restore the bounds check in the following loop to catch this.)

carlobaldassi commented on base/bitarray.jl:
```diff
@@ -1377,14 +1377,90 @@ end
+
+    chunk_start = _div64(start)+1
+    within_chunk_start = _mod64(start)
+    mask = ~(_msk64 << within_chunk_start)
```
should be `mask = _msk_end(start)` I think.

mbauman replied:
Ah, right you are.  That's a much more straightforward way of thinking through the logic here.  My thought process here was much more contorted.

carlobaldassi commented on base/bitarray.jl:
```diff
@@ -1377,14 +1377,90 @@ end
+
+    chunk_start = _div64(start)+1
+    within_chunk_start = _mod64(start)
+    mask = _msk64 << within_chunk_start
```
I think these lines should be

```
chunk_start = _div64(start-1) + 1
mask = ~_msk_end(start)
```

carlobaldassi commented on base/bitarray.jl:
```diff
@@ -1377,14 +1377,90 @@ end
+
+    start > 0 || return 0
+    start > length(B) && throw(BoundsError(B, start))
+    length(B) == 0 && return 0
```
same as above

carlobaldassi commented:
Thanks! I left some notes, besides that all the testing is certainly most needed and welcome!
I suggest, however, when testing BitArray stuff, to test with particular fastidiousness the cases around the chunks borders, e.g. using BitArrays of length 0,1,63,64,65,127, you get the idea. These are always the trickiest to get right.

Also thanks for catching the bug in `findnext`, it was a leftover from when it was just called `findfirst` which got unnoticed in the switch, I think.

mbauman commented:
Thank you for the review, @carlobaldassi!  I have addressed all comments with my updated commits.  I can rebase and squash them back into two commits again, but I think once I do that we lose your inline notes on the previous commits.

carlobaldassi commented:
Ok this looks good to me. I'm for squashing them back into two clean commits, inline comments are not really important and besides I think github keeps them anyway. Then whenever it turns green it is ok to merge for me.

mbauman commented:
Rebased. Travis passed; Appveyor seems to be having some infrastructure troubles today.

tkelman commented:
@carlobaldassi what about backporting just the new tests from a1289ec4fbb5c4b643fc0dd6734f10618497a829, then 37040f70a9267b7b38d59793fb01fff97c6c5e9f in its entirety? Assuming that passes locally, that is.

mbauman commented:
Yeah, I should have ordered these commits the other way around.  Sorry about that.  The only relevant commits to the back port are the ones I uncomment in that commit.  I can get to this next week myself, but I'm afraid I won't have time until then.

tkelman commented:
Thanks for manually doing the backport, moving the tracking label to #9833. Will look at in more detail shortly.
